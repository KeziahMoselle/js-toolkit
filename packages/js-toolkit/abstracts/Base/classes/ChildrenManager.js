/**
 * @typedef {import('./index.js').default} Base
 * @typedef {import('./index.js').BaseComponent} BaseComponent
 * @typedef {import('./index.js').BaseAsyncComponent} BaseAsyncComponent
 * @typedef {import('./index.js').BaseConfigComponents} BaseConfigComponents
 */

/**
 * Get a list of elements based on the name of a component.
 *
 * @param {String} nameOrSelector
 *   The name or selector to used for this component.
 * @param {HTMLElement|Document} element
 *   The root element on which to query the selector, defaults to `document`.
 * @return {Array<HTMLElement>}
 *   A list of elements on which the component should be mounted.
 */
export function getComponentElements(nameOrSelector, element = document) {
  const selector = `[data-component="${nameOrSelector}"]`;
  let elements = [];

  try {
    elements = Array.from(element.querySelectorAll(selector));
    // eslint-disable-next-line no-empty
  } catch {}

  // If no child component found with the default selector, try a classic DOM selector
  if (elements.length === 0) {
    elements = Array.from(element.querySelectorAll(nameOrSelector));
  }

  return elements;
}

/**
 * Children manager.
 */
export default class ChildrenManager {
  /** @type {Base} */
  #base;

  /** @type {HTMLElement} */
  #element;

  /** @type {BaseConfigComponents} */
  #components;

  /**
   * @return {string[]}
   */
  get #registeredNames() {
    return Object.keys(this);
  }

  /**
   * Class constructor.
   * @param  {Base}                 instance   The component's instance.
   * @param  {HTMLElement}          element    The component's root element
   * @param  {BaseConfigComponents} components The children components' classes
   */
  constructor(base, element, components) {
    this.#base = base;
    this.#element = element;
    this.#components = components;
  }

  /**
   * Register instances of all children components.
   */
  registerAll() {
    Object.entries(this.#components).forEach(([name, component]) =>
      this.#register(name, component)
    );
  }

  /**
   * Register instance of a child component.
   *
   * @param {string} name
   *   The name of the child component.
   * @param {BaseComponent|BaseAsyncComponent} component
   *   A Base class or a Promise for async components.
   */
  #register(name, component) {
    Object.defineProperty(this, name, {
      enumerable: true,
      configurable: true,
      get: () => {
        const elements = getComponentElements(name, this.#element);

        if (elements.length === 0) {
          return [];
        }

        return elements
          .map((element) => this.#getChild(element, component))
          .filter((instance) => instance !== 'terminated');
      },
    });
  }

  /**
   * Get a child component's instance.
   *
   * @param {HTMLElement & { __base__?: Base | 'terminated' }} el
   *   The root element of the child component.
   * @param {BaseComponent|BaseAsyncComponent} ComponentClass
   *   A Base class or a Promise for async components.
   * @return {Base|Promise|'terminated'}
   *   A Base instance or a Promise resolving to a Base instance.
   */
  #getChild(el, ComponentClass) {
    // Return existing instance if it exists
    if (el.__base__) {
      return el.__base__;
    }

    // Return a new instance if the component class is a child of the Base class
    if ('$isBase' in ComponentClass) {
      const child = new ComponentClass(el);
      Object.defineProperty(child, '$parent', { get: () => this.#base });
      return child;
    }

    // Resolve async components
    return ComponentClass().then((module) => {
      return this.#getChild(el, module.default ?? module);
    });
  }

  /**
   * Mount all child component instances.
   */
  mountAll() {
    this.#triggerHookForAll('$mount');
  }

  /**
   * Update all child component instances.
   */
  updateAll() {
    this.#triggerHookForAll('$update');
  }

  /**
   * Destroy all child component instances.
   */
  destroyAll() {
    this.#triggerHookForAll('$destroy');
  }

  /**
   * Execute the given hook for all children instances.
   *
   * @param {'$mount'|'$update'|'$destroy'} hook The hook to execute.
   */
  #triggerHookForAll(hook) {
    this.#registeredNames.forEach((name) => {
      this[name].forEach((instance) => {
        if (instance instanceof Promise) {
          instance.then((resolvedInstance) => this.#triggerHook(hook, resolvedInstance));
        } else {
          this.#triggerHook(hook, instance);
        }
      });
    });
  }

  /**
   * Execute the given hook for the given instance.
   *
   * @param {'$mount'|'$update'|'$destroy'} hook The hook to trigger.
   * @param {Base} instance The target instance.
   */
  #triggerHook(hook, instance) {
    if (hook === '$update' && !instance.$isMounted) {
      // eslint-disable-next-line no-param-reassign
      hook = '$mount';
    }
    instance[hook]();
  }
}
