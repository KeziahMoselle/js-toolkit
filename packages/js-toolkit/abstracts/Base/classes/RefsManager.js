import { warn } from '../utils.js';

/**
 * @typedef {import('../index.js').default} Base
 */

/**
 * Refs Manager
 */
export default class RefsManager {
  #element;

  #refs;

  #base;

  /**
   * Class constructor.
   *
   * @param {Base}        base    The base instance tied to this manager.
   * @param {HTMLElement} element The root element of a Base component.
   * @param {string[]}    refs    The list of refs of this component.
   */
  constructor(base, element, refs) {
    this.#element = element;
    this.#refs = refs;
    this.#base = base;
  }

  /**
   * Register all refs.
   */
  registerAll() {
    this.#refs.forEach((refName) => this.#register(refName));
    this.#base.$emit('get:refs', this);
  }

  /**
   * Register one ref.
   *
   * @param {string} refName The ref name.
   */
  #register(refName) {
    const isMultiple = refName.endsWith('[]');
    const propName = isMultiple ? refName.replace(/\[\]$/, '') : refName;

    const refs = Array.from(this.#element.querySelectorAll(`[data-ref="${refName}"]`))
      .filter(
        /**
         * @param {HTMLElement} ref
         */
        (ref) => this.#filterRefsBelongingToInstance(ref)
      )
      .map(
        /**
         * @param {HTMLElement & { __base__?: Base }} ref
         */
        (ref) => ref.__base__ ?? ref
      );

    if (!isMultiple && refs.length > 1) {
      warn(
        this.#base,
        `The "${refName}" ref has been found multiple times.`,
        'Did you forgot to add the `[]` suffix to its name?'
      );
    }

    Object.defineProperty(this, propName, {
      // @todo remove usage of non-multiple refs as array
      value: isMultiple || refs.length > 1 ? refs : refs[0],
      enumerable: true,
    });
  }

  /**
   * Filter refs belonging to the related Base instance.
   * @param {HTMLElement} ref The ref to test.
   */
  #filterRefsBelongingToInstance(ref) {
    let ancestor = ref.parentElement;

    while (ancestor && ancestor.dataset.component === undefined) {
      ancestor = ancestor.parentElement;
    }

    return ancestor === null || ancestor === this.#element;
  }
}
