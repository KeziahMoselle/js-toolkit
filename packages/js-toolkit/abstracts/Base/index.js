import autoBind from '../../utils/object/autoBind.js';
import EventManager from '../EventManager.js';
import { callMethod, debug, getConfig, getComponentElements } from './utils.js';
import { getChildren } from './children.js';
import { getOptions } from './options.js';
import { getRefs } from './refs.js';
import { mountComponents, mountOrUpdateComponents, destroyComponents } from './components.js';
import Services from './classes/Services.js';
import bindEvents from './events.js';

// Define the __DEV__ constant if not defined
if (typeof __DEV__ === 'undefined') {
  if (typeof globalThis !== 'undefined') {
    globalThis.__DEV__ = false;
  } else if (typeof window !== 'undefined') {
    window.__DEV__ = false;
  }
}

let id = 0;

/**
 * @typedef {typeof Base} BaseComponent
 * @typedef {(Base) => Promise<BaseComponent | { default: BaseComponent }>} BaseAsyncComponent
 * @typedef {{ name: string, debug: boolean, log: boolean, [name:string]: any }} BaseOptions
 * @typedef {{ [name:string]: HTMLElement | HTMLElement[] | Base | Base[] | Promise<Base> | Promise<Base>[] }} BaseRefs
 * @typedef {{ [nameOrSelector:string]: Base[] | Promise<Base>[] }} BaseChildren
 * @typedef {{ [nameOrSelector:string]: BaseComponent | BaseAsyncComponent }} BaseConfigComponents
 * @typedef {import('./classes/Options').OptionsSchema} BaseConfigOptions
 */

/**
 * @typedef {Object} BaseConfig
 * @property {String} name
 * @property {Boolean} [debug]
 * @property {Boolean} [log]
 * @property {String[]} [refs]
 * @property {BaseConfigComponents} [components]
 * @property {BaseConfigOptions} [options]
 */

/**
 * @typedef {import('./classes/Services.js').ServiceName} ServiceName
 */

/**
 * Base class to easily create components.
 *
 * @example
 * ```js
 * class Component extends Base {
 *   static config = {
 *     name: 'Component',
 *     log: true,
 *   };
 *
 *   mounted() {
 *     this.$log('Component is mounted!');
 *   }
 * }
 *
 * class App extends Base {
 *   static config = {
 *     name: 'App',
 *     components: {
 *       Component,
 *     },
 *   };
 * }
 *
 * new App(document.body).$mount();
 * ```
 */
export default class Base extends EventManager {
  /**
   * The instance parent.
   * @type {Base}
   */
  $parent = null;

  /**
   * The instance id.
   * @type {string}
   */
  $id;

  /**
   * The root element.
   * @type {HTMLElement}
   */
  $el;

  /**
   * The instance options.
   * @type {BaseOptions}
   */
  $options;

  /**
   * The instance children components.
   * @type {BaseChildren}
   */
  $children;

  /**
   * The instance services.
   * @type {Services}
   */
  $services;

  /**
   * The state of the component.
   * @type {Boolean}
   */
  $isMounted = false;

  /**
   * This is a Base instance.
   * @type {Boolean}
   */
  static $isBase = true;

  /**
   * Get properties to exclude from the autobind call.
   * @return {Array<String|RegExp>}
   */
  get _excludeFromAutoBind() {
    return [
      '$mount',
      '$update',
      '$destroy',
      '$terminate',
      '$log',
      '$on',
      '$once',
      '$off',
      '$emit',
      'mounted',
      'loaded',
      'ticked',
      'resized',
      'moved',
      'keyed',
      'scrolled',
      'destroyed',
      'terminated',
    ];
  }

  /**
   * @deprecated Use the static `config` property instead.
   * @return {BaseConfig}
   */
  get config() {
    return null;
  }

  /** @type {BaseConfig} */
  static config;

  /**
   * Get the component's refs.
   * @return {BaseRefs}
   */
  get $refs() {
    return getRefs(this, this.$el);
  }

  /**
   * Class constructor where all the magic takes place.
   *
   * @param {HTMLElement} element The component's root element dd.
   */
  constructor(element) {
    super();

    if (!element) {
      throw new Error('The root element must be defined.');
    }

    const { name } = getConfig(this);

    /** @type {String} */
    this.$id = `${name}-${id}`;
    id += 1;

    /** @type {HTMLElement} */
    this.$el = element;

    this.$options = getOptions(this, element, getConfig(this));
    this.$children = getChildren(this, this.$el, getConfig(this).components || {});
    this.$services = new Services(this);

    if (!('__base__' in this.$el)) {
      Object.defineProperty(this.$el, '__base__', {
        get: () => this,
        configurable: true,
      });
    }

    // Autobind all methods to the instance
    autoBind(this, {
      exclude: [...this._excludeFromAutoBind],
    });

    let unbindMethods = [];
    this.$on('mounted', () => {
      this.$services.enableAll();
      unbindMethods = [...bindEvents(this)];
      mountComponents(this);
      this.$isMounted = true;
    });

    this.$on('updated', () => {
      this.$services.disableAll();
      unbindMethods.forEach((method) => method());
      unbindMethods = [...bindEvents(this)];
      this.$services.enableAll();
      mountOrUpdateComponents(this);
    });

    this.$on('destroyed', () => {
      this.$isMounted = false;
      this.$services.disableAll();
      unbindMethods.forEach((method) => method());
      destroyComponents(this);
    });

    if (__DEV__) {
      debug(this, 'constructor', this);
    }
    return this;
  }

  /**
   * Small helper to log stuff.
   *
   * @return {(...args: any) => void} A log function if the log options is active.
   */
  get $log() {
    return this.$options.log
      ? window.console.log.bind(window, `[${this.$options.name}]`)
      : function noop() {};
  }

  /**
   * Trigger the `mounted` callback.
   */
  $mount() {
    if (__DEV__) {
      debug(this, '$mount');
    }

    callMethod(this, 'mounted');
    return this;
  }

  /**
   * Update the instance children.
   */
  $update() {
    if (__DEV__) {
      debug(this, '$update');
    }

    callMethod(this, 'updated');
    return this;
  }

  /**
   * Trigger the `destroyed` callback.
   */
  $destroy() {
    if (__DEV__) {
      debug(this, '$destroy');
    }

    callMethod(this, 'destroyed');
    return this;
  }

  /**
   * Terminate a child instance when it is not needed anymore.
   * @return {void}
   */
  $terminate() {
    if (__DEV__) {
      debug(this, '$terminate');
    }

    // First, destroy the component.
    this.$destroy();

    // Execute the `terminated` hook if it exists
    callMethod(this, 'terminated');

    // Delete the reference to the instance
    // delete this.$el.__base__;

    // And update its status to prevent re-instantiation when accessing the
    // parent's `$children` property
    Object.defineProperty(this.$el, '__base__', {
      value: 'terminated',
      configurable: false,
      writable: false,
    });
  }

  /**
   * Factory method to generate multiple instance of the class.
   *
   * @param  {String}      nameOrSelector The selector on which to mount each instance.
   * @return {Array<Base>}                A list of the created instance.
   */
  static $factory(nameOrSelector) {
    if (!nameOrSelector) {
      throw new Error(
        'The $factory method requires a componentâ€™s name or selector to be specified.'
      );
    }

    return getComponentElements(nameOrSelector).map((el) => new this(el).$mount());
  }
}
