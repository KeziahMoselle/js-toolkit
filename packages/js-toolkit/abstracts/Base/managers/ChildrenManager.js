import { getComponentElements } from '../utils.js';

/**
 * @typedef {import('../index.js').default} Base
 * @typedef {import('../index.js').BaseComponent} BaseComponent
 * @typedef {import('../index.js').BaseAsyncComponent} BaseAsyncComponent
 * @typedef {import('../index.js').BaseConfigComponents} BaseConfigComponents
 */

/**
 * Children manager.
 */
export default class ChildrenManager {
  /** @type {Base} */
  #base;

  /** @type {HTMLElement} */
  #element;

  /** @type {BaseConfigComponents} */
  #components;

  /**
   * @return {string[]}
   */
  get #registeredNames() {
    return Object.keys(this);
  }

  /**
   * Class constructor.
   * @param  {Base}                 base       The component's instance.
   * @param  {HTMLElement}          element    The component's root element
   * @param  {BaseConfigComponents} components The children components' classes
   */
  constructor(base, element, components) {
    this.#base = base;
    this.#element = element;
    this.#components = components;
  }

  /**
   * Register instances of all children components.
   */
  registerAll() {
    Object.entries(this.#components).forEach(([name, component]) =>
      this.#register(name, component)
    );
  }

  /**
   * Register instance of a child component.
   *
   * @param {string} name
   *   The name of the child component.
   * @param {BaseComponent|BaseAsyncComponent} component
   *   A Base class or a Promise for async components.
   */
  #register(name, component) {
    Object.defineProperty(this, name, {
      enumerable: true,
      configurable: true,
      get: () => {
        const elements = getComponentElements(name, this.#element);

        if (elements.length === 0) {
          return [];
        }

        return elements
          .map((element) => this.#getChild(element, component))
          .filter((instance) => instance !== 'terminated');
      },
    });
  }

  /**
   * Get a child component's instance.
   *
   * @param {HTMLElement & { __base__?: Base | 'terminated' }} el
   *   The root element of the child component.
   * @param {BaseComponent|BaseAsyncComponent} ComponentClass
   *   A Base class or a Promise for async components.
   * @return {Base|Promise|'terminated'}
   *   A Base instance or a Promise resolving to a Base instance.
   */
  #getChild(el, ComponentClass) {
    // Return existing instance if it exists
    if (el.__base__) {
      return el.__base__;
    }

    // Return a new instance if the component class is a child of the Base class
    if ('$isBase' in ComponentClass) {
      const child = new ComponentClass(el);
      Object.defineProperty(child, '$parent', { get: () => this.#base });
      return child;
    }

    // Resolve async components
    return ComponentClass(this.#base).then((module) => {
      // @ts-ignore
      return this.#getChild(el, module.default ?? module);
    });
  }

  /**
   * Mount all child component instances.
   */
  mountAll() {
    this.#triggerHookForAll('$mount');
  }

  /**
   * Update all child component instances.
   */
  updateAll() {
    this.#triggerHookForAll('$update');
  }

  /**
   * Destroy all child component instances.
   */
  destroyAll() {
    this.#triggerHookForAll('$destroy');
  }

  /**
   * Execute the given hook for all children instances.
   *
   * @param {'$mount'|'$update'|'$destroy'} hook The hook to execute.
   */
  #triggerHookForAll(hook) {
    this.#registeredNames.forEach((name) => {
      this[name].forEach((instance) => {
        if (instance instanceof Promise) {
          instance.then((resolvedInstance) => this.#triggerHook(hook, resolvedInstance));
        } else {
          this.#triggerHook(hook, instance);
        }
      });
    });
  }

  /**
   * Execute the given hook for the given instance.
   *
   * @param {'$mount'|'$update'|'$destroy'} hook The hook to trigger.
   * @param {Base} instance The target instance.
   */
  #triggerHook(hook, instance) {
    if (hook === '$update' && !instance.$isMounted) {
      // eslint-disable-next-line no-param-reassign
      hook = '$mount';
    }
    instance[hook]();
  }
}
