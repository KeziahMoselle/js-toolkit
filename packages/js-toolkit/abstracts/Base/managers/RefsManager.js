/**
 * @typedef {import('../index.js').default} Base
 * @typedef {import('./EventsManager.js').default} EventsManager
 */

/**
 * Refs Manager
 */
export default class RefsManager {
  /**
   * @type {HTMLElement}
   * @private
   */
  __element;

  /**
   * @type {string[]}
   * @private
   */
  __refs;

  /**
   * @type {Base}
   * @private
   */
  __base;

  /**
   * @type {EventsManager}
   * @private
   */
  __eventsManager;

  /**
   * Class constructor.
   *
   * @param {Base} base
   *   The base instance tied to this manager.
   * @param {HTMLElement} element
   *   The root element of a Base component.
   * @param {string[]} refs
   *   The list of refs of this component.
   * @param {EventsManager} eventsManager
   */
  constructor(base, element, refs, eventsManager) {
    this.__element = element;
    this.__refs = refs;
    this.__base = base;
    this.__eventsManager = eventsManager;
  }

  /**
   * Register all refs.
   *
   * @return {void}
   */
  registerAll() {
    this.__refs.forEach((refName) => this.__register(refName));
  }

  /**
   * Unregister all refs.
   *
   * @return {void}
   */
  unregisterAll() {
    this.__refs.forEach((refName) => this.__unregister(refName));
  }

  /**
   * Register one ref.
   *
   * @param {string} refName The ref name.
   * @private
   */
  __register(refName) {
    const isMultiple = refName.endsWith('[]');
    const propName = this.__normalizeRefName(refName);

    const refs = /** @type {HTMLElement[]} */ (
      Array.from(this.__element.querySelectorAll(`[data-ref="${refName}"]`))
    ).filter(
      /**
       * @param {HTMLElement} ref
       */
      (ref) => this.__filterRefsBelongingToInstance(ref)
    );

    if (!isMultiple && refs.length > 1) {
      console.warn(
        `[${this.__base.$options.name}]`,
        `The "${refName}" ref has been found multiple times.`,
        'Did you forgot to add the `[]` suffix to its name?'
      );
    }

    this.__eventsManager.bindRef(refName, refs);

    Object.defineProperty(this, propName, {
      // @todo remove usage of non-multiple refs as array
      value: isMultiple || refs.length > 1 ? refs : refs[0],
      enumerable: true,
      configurable: true,
    });
  }

  /**
   * Unregister one ref.
   *
   * @param {string} refName The ref name.
   * @private
   */
  __unregister(refName) {
    const propName = this.__normalizeRefName(refName);
    const refs = Array.isArray(this[propName]) ? this[propName] : [this[propName]];
    this.__eventsManager.unbindRef(refName, refs);
  }

  /**
   * Normalize the name of ref.
   *
   * @param {string} name The original name.
   * @return {string}     The normalized name.
   * @private
   */
  __normalizeRefName(name) {
    return name.endsWith('[]') ? name.replace(/\[\]$/, '') : name;
  }

  /**
   * Filter refs belonging to the related Base instance.
   *
   * @param {HTMLElement} ref The ref to test.
   * @return {boolean}
   * @private
   */
  __filterRefsBelongingToInstance(ref) {
    let ancestor = ref.parentElement;

    while (ancestor && ancestor.dataset.component === undefined) {
      ancestor = ancestor.parentElement;
    }

    return ancestor === null || ancestor === this.__element;
  }
}
