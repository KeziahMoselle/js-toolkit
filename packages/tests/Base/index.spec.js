/* eslint-disable no-new, require-jsdoc, max-classes-per-file */
import { jest } from '@jest/globals';
import Base from '@studiometa/js-toolkit';
import ChildrenManager from '@studiometa/js-toolkit/Base/managers/ChildrenManager';
import OptionsManager from '@studiometa/js-toolkit/Base/managers/OptionsManager';
import RefsManager from '@studiometa/js-toolkit/Base/managers/RefsManager';
import wait from '../__utils__/wait';

describe('The abstract Base class', () => {
  it('must be extended', () => {
    expect(() => {
      new Base(document.createElement('div'));
    }).toThrow();
  });

  it('should throw an error when extended without proper configuration', () => {
    expect(() => {
      class Foo extends Base {}
      new Foo(document.createElement('div'));
    }).toThrow('The `config` static property must be defined.');

    expect(() => {
      // @ts-ignore
      class Foo extends Base {
        static config = {};
      }
      new Foo(document.createElement('div'));
    }).toThrow('The `config.name` property is required.');
  });

  it('should throw an error if instantiated without a root element.', () => {
    expect(() => {
      class Foo extends Base {
        static config = {
          name: 'Foo',
        };
      }
      // @ts-ignore
      new Foo();
    }).toThrow('The root element must be defined.');
  });
});

describe('A Base instance', () => {
  class Foo extends Base {
    static config = {
      name: 'Foo',
    };
  }
  const element = document.createElement('div');
  const foo = new Foo(element).$mount();

  it('should have an `$id` property', () => {
    expect(foo.$id).toBeDefined();
  });

  it('should have an `$isMounted` property', () => {
    expect(foo.$isMounted).toBe(true);
  });

  it('should have a `$refs` property', () => {
    expect(foo.$refs).toBeInstanceOf(RefsManager);
  });

  it('should have a `$children` property', () => {
    expect(foo.$children).toBeInstanceOf(ChildrenManager);
  });

  it('should have an `$options` property', () => {
    expect(foo.$options).toBeInstanceOf(OptionsManager);
    expect(foo.$options.name).toBe('Foo');
  });

  it('should be able to set any `$options` property', () => {
    foo.$options.log = true;
    expect(foo.$options.log).toBe(true);
  });

  it('should have an `$el` property', () => {
    expect(foo.$el).toBe(element);
  });

  it('should have a `__base__` property', () => {
    // @ts-ignore
    expect(foo.$el.__base__).toBe(foo);
  });

  it('should inherit from parent config', () => {
    class A extends Base {
      static config = {
        name: 'A',
        log: true,
      };
    }

    class B extends A {
      static config = {
        name: 'B',
        options: {
          title: String,
          color: String,
        },
      };
    }

    class C extends B {
      static config = {
        name: 'C',
        options: {
          color: Boolean,
        },
      };
    }

    class D extends C {
      static config = {
        name: 'D',
        log: false,
      };
    }

    const d = new D(document.createElement('div'));
    expect(d.__config).toEqual({
      name: 'D',
      log: false,
      options: {
        color: Boolean,
        title: String,
      },
    });
  });
});

describe('A Base instance methods', () => {
  class Foo extends Base {
    static config = {
      name: 'Foo',
    };
  }

  let element;
  let foo;

  beforeEach(() => {
    element = document.createElement('div');
    foo = new Foo(element).$mount();
  });

  it('should emit a mounted event', () => {
    const fn = jest.fn();
    foo.$on('mounted', fn);
    foo.$destroy();
    foo.$mount();
    expect(fn).toHaveBeenCalledTimes(1);
    foo.$mount();
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should emit a destroyed event', () => {
    const fn = jest.fn();
    foo.$on('destroyed', fn);
    foo.$destroy();
    expect(fn).toHaveBeenCalledTimes(1);
    foo.$destroy();
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should be able to update its child components.', async () => {
    const div = document.createElement('div');
    div.innerHTML = `
      <div data-component="Bar"></div>
      <div data-component="Bar"></div>
    `;

    const fn = jest.fn();
    class Bar extends Foo {}
    class Baz extends Base {
      static config = { name: 'Baz' };
      updated() {
        fn();
      }
    }

    class App extends Base {
      static config = {
        name: 'App',
        components: {
          Bar,
          Baz,
          AsyncBaz: () => Promise.resolve(Baz),
        },
      };
    }

    const app = new App(div).$mount();
    expect(app.$children.Bar).toHaveLength(2);
    expect(app.$children.Bar[0].$isMounted).toBe(true);
    div.innerHTML = `
      <div data-component="Baz"></div>
      <div data-component="Baz"></div>
      <div data-component="AsyncBaz"></div>
    `;

    app.$update();

    expect(app.$children.Bar).toEqual([]);
    expect(app.$children.Baz).toHaveLength(2);
    expect(app.$children.Baz[0].$isMounted).toBe(true);
    const asyncBaz = await app.$children.AsyncBaz[0];
    expect(asyncBaz.$isMounted).toBe(true);

    const id = div.firstElementChild.__base__.$id;
    expect(id).toBe(app.$children.Baz[0].$id);

    const child = document.createElement('div');
    child.setAttribute('data-component', 'Baz');
    div.appendChild(child);
    expect(id).toBe(app.$children.Baz[0].$id);

    app.$update();

    expect(id).toBe(app.$children.Baz[0].$id);
    expect(fn).toHaveBeenCalledTimes(3);
  });

  it('should implement the $factory method', () => {
    class Bar extends Foo {}
    class Baz extends Foo {}
    class Boz extends Foo {}
    document.body.innerHTML = `
      <div data-component="Bar"></div>
      <div data-component="Bar"></div>
      <div class="custom-selector"></div>
      <div class="custom-selector"></div>
    `;

    const barInstances = Bar.$factory('Bar');
    const bazInstances = Baz.$factory('.custom-selector');
    const bozInstances = Boz.$factory('Boz');
    expect(barInstances).toHaveLength(2);
    expect(barInstances[0] instanceof Bar).toBe(true);
    expect(bazInstances).toHaveLength(2);
    expect(bazInstances[0] instanceof Baz).toBe(true);
    expect(bozInstances).toHaveLength(0);
    expect(Baz.$factory).toThrow(/\$factory method/);
  });

  it('should be able to be terminated', () => {
    const fn = jest.fn();
    class Bar extends Foo {
      terminated() {
        fn('method');
      }
    }

    const div = document.createElement('div');
    const bar = new Bar(div).$mount();
    expect(bar).toEqual(div.__base__);
    bar.$on('terminated', () => fn('event'));
    bar.$terminate();
    expect(fn).toHaveBeenCalledTimes(2);
    expect(fn).toHaveBeenNthCalledWith(1, 'event');
    expect(fn).toHaveBeenNthCalledWith(2, 'method');
  });

  it('should not find children if none provided', () => {
    class Bar extends Base {
      static config = {
        name: 'Bar',
      };
    }

    class Baz extends Base {
      static config = {
        name: 'Baz',
        components: { Bar },
      };
    }
    expect(foo.$children.registeredNames).toEqual([]);
    expect(new Baz(document.createElement('div')).$mount().$children.Bar).toEqual([]);
  });

  it('should not find terminated children', () => {
    class Bar extends Base {
      static config = {
        name: 'Bar',
      };
    }

    class Baz extends Base {
      static config = {
        name: 'Baz',
        components: { Bar },
      };
    }

    const div = document.createElement('div');
    div.innerHTML = `
      <div data-component="Bar"></div>
    `;
    const baz = new Baz(div).$mount();
    expect(baz.$children.Bar).toEqual([div.firstElementChild.__base__]);
    div.firstElementChild.__base__.$terminate();
    expect(baz.$children.Bar).toEqual([]);
  });

  it('should listen to the window.onload event', () => {
    const fn = jest.fn();
    class Bar extends Foo {
      loaded() {
        fn();
      }
    }

    const bar = new Bar(document.createElement('div')).$mount();
    window.dispatchEvent(new CustomEvent('load'));
    expect(fn).toHaveBeenCalledTimes(1);
    bar.$destroy();
    window.dispatchEvent(new CustomEvent('load'));
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should mount and destroy its children', () => {
    class Bar extends Base {
      static config = {
        name: 'Bar',
      };
    }
    class Baz extends Foo {
      static config = {
        name: 'Baz',
        components: { Bar },
      };
    }

    document.body.innerHTML = `<div data-component="Bar"></div>`;
    const baz = new Baz(document.body).$mount();
    const barElement = document.querySelector('[data-component="Bar"]');
    expect(baz.$isMounted).toBe(true);
    expect(barElement.__base__.$isMounted).toBe(true);
    baz.$destroy();
    expect(baz.$isMounted).toBe(false);
    expect(barElement.__base__.$isMounted).toBe(false);
  });
});

describe('A Base instance config', () => {
  let element;

  beforeEach(() => {
    element = document.createElement('div');
  });

  it('should have a working $log method when active', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        log: true,
      };
    }
    const spy = jest.spyOn(window.console, 'log');
    spy.mockImplementation(() => true);
    const foo = new Foo(element).$mount();
    expect(foo.$options.log).toBe(true);
    foo.$log('bar');
    expect(spy).toHaveBeenCalledWith('[Foo]', 'bar');
    spy.mockRestore();
  });

  it('should have a silent $log method when not active', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        log: false,
      };
    }
    const spy = jest.spyOn(window.console, 'log');
    const foo = new Foo(element).$mount();
    expect(foo.$options.log).toBe(false);
    foo.$log('bar');
    expect(spy).toHaveBeenCalledTimes(0);
    spy.mockRestore();
  });

  it('should have a working debug method when active in dev mode', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        debug: true,
      };
    }
    globalThis.__DEV__ = true;
    process.env.NODE_ENV = 'development';
    const spy = jest.spyOn(window.console, 'log');
    spy.mockImplementation(() => true);
    const div = document.createElement('div');
    const foo = new Foo(div).$mount();
    expect(spy.mock.calls).toMatchSnapshot();
    spy.mockRestore();
    process.env.NODE_ENV = 'test';
  });
});
