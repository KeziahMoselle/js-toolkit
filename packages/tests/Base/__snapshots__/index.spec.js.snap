// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Base instance config should have a working debug method when active in dev mode 1`] = `
Array [
  Array [
    "[debug] [Foo-36]",
    "constructor",
    Foo {
      "$el": <div />,
      "$id": "Foo-36",
      "$isMounted": true,
      "$parent": null,
      "__children": ChildrenManager {},
      "__eventHandlers": Map {},
      "__events": EventsManager {},
      "__options": OptionsManager {
        "debug": true,
        "log": false,
        "name": "Foo",
      },
      "__refs": RefsManager {},
      "__services": ServicesManager {
        "__customServices": Object {},
      },
    },
  ],
  Array [
    "[debug] [Foo-36]",
    "$mount",
  ],
  Array [
    "[debug] [Foo-36]",
    "$emit",
    "get:children",
    Array [
      ChildrenManager {},
    ],
  ],
  Array [
    "[debug] [Foo-36]",
    "$emit",
    "get:refs",
    Array [
      RefsManager {},
    ],
  ],
  Array [
    "[debug] [Foo-36]",
    "$emit",
    "get:services",
    Array [
      ServicesManager {
        "__customServices": Object {},
      },
    ],
  ],
  Array [
    "[debug] [Foo-36]",
    "$emit",
    "get:children",
    Array [
      ChildrenManager {},
    ],
  ],
  Array [
    "[debug] [Foo-36]",
    "callMethod",
    "mounted",
  ],
  Array [
    "[debug] [Foo-36]",
    "$emit",
    "mounted",
    Array [],
  ],
]
`;
